### PR TITLE
Enforcing threat models

### LINKED ISSUE #22: Overlay threat model assumes honest perturbers
https://github.com/IntelLabs/MART/blob/6184ca764c871cc098a4e822e520f476e556a971/mart/attack/threat_model.py#L60

It should be better to return `input * (1 - mask) + perturbation * mask`, in case the perturbation is not constrained to the masked region.

### PR DESCRIPTION
**Pull request by mzweilin**
# What does this PR do?

Fixes #22

This PR enforces constraints on perturbation, in case an adversary does not comply with the threat model. MART will raise an exception if any constraint on adversarial examples is violated.

Supported constraints
- [x] Lp norm
- [x] Range
- [x] Integer
- [x] Mask 

The Lp norm constraint is added to the `input_adv_test` in `CIFAR10_CNN_Adv` experiment and the overhead is negligible. (it took the same 18 seconds to run `input_adv_test`)

I encountered round-off errors in projectors when trying to enforce the L-inf-additive threat model. Rounding the [0, 255] input to integers in data modules mitigates the round-off errors. Fixed in #112

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] tests/test_enforcer.py

## Before submitting

- [x] The title is **self-explanatory** and the description **concisely** explains the PR
- [x] My **PR does only one thing**, instead of bundling different changes together
- [ ] I list all the **breaking changes** introduced by this pull request
- [x] I have commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have run pre-commit hooks with `pre-commit run -a` command without errors

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by dxoigmn — CHANGES_REQUESTED** [review]
Can you address #22 too?

**Comment by mzweilin** [comment]
By rounding [0, 255] input to integers, we can avoid round-off errors in projectors. So I have reverted the changes below.

```diff
diff --git a/mart/attack/projector.py b/mart/attack/projector.py
index 913bce8..8da46a4 100644
--- a/mart/attack/projector.py
+++ b/mart/attack/projector.py
@@ -155,6 +155,15 @@ class LinfAdditiveRange(Projector):
         eps_min = (input - self.eps).clamp(self.min, self.max) - input
         eps_max = (input + self.eps).clamp(self.min, self.max) - input
 
+        # Avoid round-off errors.
+        eps_min = eps_min.clamp(min=-self.eps)
+        eps_max = eps_max.clamp(max=self.eps)
+
+        # TODO: We could have rounded input earlier to avoid round-off errors here.
+        # f"{torch.tensor(126.99999237060547, dtype=torch.float) + 2 - 126.99999237060547}" == "2.0000076293945312"
+        # f"{torch.tensor(126.99999237060547, dtype=torch.double) + 2 - 126.99999237060547}" == "2.0"
+        # f"{torch.tensor(126.99999237060547, dtype=torch.float).round() + 2 - torch.tensor(126.99999237060547).round()}" == "2.0"
+
         tensor.clamp_(eps_min, eps_max)
```

File: mart/attack/threat_model.py [hunk 1]
```diff
@@ -5,21 +5,60 @@
 #
 
 import abc
-from typing import Any, Dict, Union
+from typing import Any, Dict, Optional, Union
```
Comment (dxoigmn) [inline mart/attack/threat_model.py:8]: I think we should standardize on `from __future__ import annotations` and the use of `| None` for `Optional`.

File: mart/configs/datamodule/cifar10.yaml [hunk 2]
```diff
@@ -20,6 +20,13 @@ train_dataset:
       - _target_: mart.transforms.Denormalize
         center: 0
         scale: 255
+      - _target_: torch.fake_quantize_per_tensor_affine
+        _partial_: true
+        # (x/1+0).round().clamp(0, 255) * 1
+        scale: 1
+        zero_point: 0
+        quant_min: 0
+        quant_max: 255
```
Comment (dxoigmn) [inline mart/configs/datamodule/cifar10.yaml:29]: We should break this out into a new PR.

**Review by dxoigmn — COMMENTED** [review]
Do you think it would be better to integrate the enforcer into the Adversary given how verbose the current experiment configs are?

I can't actually think of a case where one would want to insert something between the perturber and model.

File: mart/attack/threat_model.py [hunk 3]
```diff
@@ -4,43 +4,122 @@
...
+        norm_max = norm_vals.max()
+        if norm_max > self.eps:
+            raise ValueError(
+                f"L-{self.p} norm of perturbation exceeds {self.eps}, reaching {norm_max}"
+            )
+
+
+class Integer(Constraint):
+    def __call__(self, input, target, input_adv):
+        torch.testing._comparison.assert_equal(input_adv, input_adv.round())
```
Comment (dxoigmn) [inline mart/attack/threat_model.py:52]: I would use `assert_close` with `rtol=0` and `atol=0` instead of the private module function here.

Comment (dxoigmn) [reply mart/attack/threat_model.py:52]: Also, this will raise an `AssertionError` which is different from every other constraint.

File: mart/attack/threat_model.py [hunk 4]
```diff
@@ -4,43 +4,122 @@
...
+class Mask(Constraint):
+    def __call__(self, input, target, input_adv):
+        # True/1 is mutable, False/0 is immutable.
+        # mask.shape=(H, W)
+        mask = target["perturbable_mask"]
+
+        # Immutable boolean mask, True is immutable.
+        imt_mask = (1 - mask).bool()
+        perturbation = input_adv - input
+        if perturbation.masked_select(imt_mask).any():
```
Comment (dxoigmn) [inline mart/attack/threat_model.py:64]: If this is slow (which is might be because masked select would seem like a slow operation), you could probably get away with using Lp(eps=0) on mask*input_adv, mask*input.

File: mart/attack/threat_model.py [hunk 5]
```diff
@@ -4,43 +4,122 @@
...
+            raise ValueError("Perturbable mask is violated.")
+
+
+class Enforcer(torch.nn.Module):
+    def __init__(self, constraints=None) -> None:
+        super().__init__()
+        self.constraints = constraints
+
+    def _check_constraints(self, input, target, input_adv):
+        if self.constraints is not None:
```
Comment (dxoigmn) [inline mart/attack/threat_model.py:74]: Fail fast.
```python
if self.constraints it None:
    return
```

Comment (dxoigmn) [reply mart/attack/threat_model.py:74]: I would actually change Line 71 to:
```python
self.constraints = constraints or []
```

and get rid of the conditional check.

File: mart/attack/threat_model.py [hunk 6]
```diff
@@ -4,43 +4,122 @@
...
+        mask = target["perturbable_mask"]
+
+        # Immutable boolean mask, True is immutable.
+        imt_mask = (1 - mask).bool()
+        perturbation = input_adv - input
+        if perturbation.masked_select(imt_mask).any():
+            raise ValueError("Perturbable mask is violated.")
+
+
+class Enforcer(torch.nn.Module):
```
Comment (dxoigmn) [inline mart/attack/threat_model.py:68]: Does this need to be a Module?

File: mart/attack/adversary.py [hunk 7]
```diff
@@ -317,6 +319,10 @@ def forward(
         # Get perturbation and apply threat model
         # The mask projector in perturber may require information from target.
         perturbation = self.perturber(input, target)
-        output = self.threat_model(input, target, perturbation)
+        output = self.composer(input, target, perturbation)
+
+        if model is not None:
+            # We only enforce constraints after the attack optimization ends.
+            self.enforcer(input, target, output)
```
[the remark this answers is not in the document]
Comment (dxoigmn) [reply mart/attack/adversary.py:326]: If you do this, you probably don't have to change the configs so much too.

**Review by dxoigmn — CHANGES_REQUESTED** [review]
Some comments on the code.

File: mart/attack/threat_model.py (hunk 5 above)
Comment (mzweilin) [reply mart/attack/threat_model.py]: Fixed.

File: mart/attack/threat_model.py [hunk 8]
```diff
@@ -4,43 +4,122 @@
...
 
 import torch
 
-__all__ = ["BatchThreatModel"]
+__all__ = ["BatchComposer"]
+
+
+class Constraint(abc.ABC):
+    @abc.abstractclassmethod
+    def __call__(self, input, target, input_adv) -> None:
```
[the remark this answers is not in the document]
Comment (mzweilin) [reply mart/attack/threat_model.py:19]: `input_adv` is moved to the first place.

I am not making `input` and `target` optional because all constraints are fed with the same arguments in a for loop.

File: mart/attack/adversary.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mzweilin) [reply mart/attack/adversary.py:326]: Running the enforcer at every attack iteration may introduce unnecessary overhead. We only enforce constraints on the final output when we run an external adversary.

File: mart/attack/adversary_wrapper.py [hunk 9]
```diff
@@ -48,5 +53,10 @@ def model_wrapper(x):
             return logits
 
         attack = self.adversary(model_wrapper)
+        input_adv = attack(input / 255, target)
+
+        # Round to integer, in case of imprecise scaling.
+        input_adv = (input_adv * 255).round()
+        self.enforcer(input, target, input_adv)
```
[the remark this answers is not in the document]
Comment (mzweilin) [reply mart/attack/adversary_wrapper.py]: I think it can stay here in the wrapper, because it is always configurable.

File: tests/test_enforcer.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+def test_constraint_integer():
+    input, target = None, None
+
+    constraint = Integer()
+
+    input_adv = torch.tensor([1.0, 2.0])
+    constraint(input_adv, input, target)
+
+    input_adv = torch.tensor([1.0, 2.001])
+    with pytest.raises(Exception):
```
Comment (dxoigmn) [inline tests/test_enforcer.py:55]: `ValueError `

File: tests/test_enforcer.py [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+def test_constraint_mask():
+    input = torch.zeros((3, 2, 2))
+    perturbation = torch.ones((3, 2, 2))
+    mask = torch.tensor([[0.0, 1.0], [1.0, 0.0]])
+    target = {"perturbable_mask": mask}
+
+    constraint = Mask()
+
+    constraint(input + perturbation * mask, input, target)
+    with pytest.raises(Exception):
```
Comment (dxoigmn) [inline tests/test_enforcer.py:68]: `ValueError `

File: tests/test_enforcer.py [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+def test_constraint_range():
+    input = torch.tensor([0, 0, 0])
+    target = None
+
+    constraint = Range(min=0, max=255)
+
+    perturbation = torch.tensor([0, 128, 255])
+    constraint(input + perturbation, input, target)
+
+    with pytest.raises(Exception):
```
[the remark this answers is not in the document]
Comment (dxoigmn) [reply tests/test_enforcer.py:22]: It might even be worth creating you own exception so you know the type of exception is intentional. That is, can anything else raise a `ValueError`?

**Review by dxoigmn — CHANGES_REQUESTED** [review]
Some comments on the tests.

File: mart/attack/enforcer.py [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+        mask = target["perturbable_mask"]
+
+        # Immutable boolean mask, True is immutable.
+        imt_mask = (1 - mask).bool()
+        perturbation = input_adv - input
+        if perturbation.masked_select(imt_mask).any():
+            raise ValueError("Perturbable mask is violated.")
+
+
+class Enforcer(torch.nn.Module):
```
Comment (dxoigmn) [inline mart/attack/enforcer.py:74]: I would not make this an `nn.Module`.

File: mart/attack/threat_model.py (hunk 4 above)
Comment (dxoigmn) [reply mart/attack/threat_model.py]: Hmm okay.

Comment (dxoigmn) [reply mart/attack/threat_model.py]: I don't quite understand but there is probably some subtly I'm missing.

File: mart/attack/threat_model.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mzweilin) [reply mart/attack/threat_model.py]: Thanks for the suggestion. `input` and `target` are made keyword arguments.

File: tests/test_adversary.py [hunk 14]
```diff
@@ -46,7 +52,12 @@ def test_adversary_with_model(input_data, target_data, perturbation):
     gain = Mock(return_value=torch.tensor(0.0, requires_grad=True))
 
     adversary = Adversary(
-        threat_model=threat_model, perturber=perturber, optimizer=optimizer, max_iters=3, gain=gain
+        composer=composer,
+        enforcer=enforcer,
```
[the remark this answers is not in the document]
Comment (mzweilin) [reply tests/test_adversary.py:56]: Added.

File: tests/test_enforcer.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (mzweilin) [reply tests/test_enforcer.py]: Thanks for the suggestion. A customized exception `ConstraintViolated` is added.

File: mart/attack/enforcer.py (hunk 13 above)
Comment (mzweilin) [reply mart/attack/enforcer.py]: Changed.

File: mart/attack/threat_model.py (hunk 4 above)
Comment (mzweilin) [reply mart/attack/threat_model.py]: You are right. I have simplified the implementation.

**Review by dxoigmn — APPROVED** [review]
LGTM! Bye bye `ThreatModel` :)
